test/langtools/jdk/jshell/TerminalNoExecTest.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8334433
+ * @summary Verify that when running JShell on platforms that support FFMTerminalProvider,
+ *          no new processes are spawned.
+ * @requires os.family == "windows" | os.family == "mac" | os.family == "linux"
+ * @library /tools/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.main
+ *          jdk.compiler/com.sun.tools.javac.util
+ * @build toolbox.ToolBox toolbox.JavaTask TerminalNoExecTest
+ * @run main TerminalNoExecTest
+ */
+
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import jdk.jfr.consumer.RecordingStream;
+import jdk.jshell.tool.JavaShellToolBuilder;
+
+import toolbox.ToolBox;
+
+public class TerminalNoExecTest {
+
+    public static void main(String... args) throws Exception {
+        if (args.length > 0) {
+            AtomicBoolean spawnedNewProcess = new AtomicBoolean();
+            try (var rs = new RecordingStream()) {
+                rs.enable("jdk.ProcessStart").withoutThreshold();
+                rs.onEvent(evt -> {
+                    System.err.println("evt: " + evt);
+                    spawnedNewProcess.set(true);
+                });
+                rs.startAsync();
+                JavaShellToolBuilder.builder().run("--execution=local", "--no-startup");
+                rs.stop();
+            }
+            if (spawnedNewProcess.get()) {
+                System.err.println("Spawned a new process!");
+                System.exit(1);
+            }
+            System.exit(0);
```

Comment:
Hello Jan, calling `System.exit()` from within the jtreg test isn't recommended as noted in https://openjdk.org/jtreg/faq.html#should-a-test-call-the-system.exit-method. Just throwing a exception should fail the test here - something like `throw new AssertionError("jshell unexpectedly spawned a new process");`

Reply:
Please ignore my previous comment. I now see what's being done here - although this is a jtreg test class, the test itself launches this class as a main application through `ProcessBuilder` and it's only in that flow that the `System.exit(...)` gets called. I think that is fine.

Reply:
FWIW, the reason to start another process is that we need (more) control over the stdin/stdout here, which is not (I think) possible using jtreg. Hence we fork a new process, that performs the real test. The test is still somewhat fragile.

Reply:
Launching of a new `java` application from within the jtreg test is fine and is done in several places. What confused me was that the test was launching the test class itself as an application. I don't know if it's a convention, but I usually have a separate nested `public class FooBarWhatever`  within the test class which I launch using the ProcessBuilder as `Test$FooBarWhatever` and have the application logic reside in the main() method of `FooBarWhatever`. 

I am not expecting that you change the PR to do it that way though. Once I spent a few seconds on that code, it was clear what it was doing.